Adds literal, folded, and single-quote scalar dump

isaacs opened the pull request:
Expands the ways that string values can be dumped to yaml.

Use simple scalar when possible, as before.

If simple is not possible, but single-quoted is allowed (ie, no
unprintables or escaped characters), then use that.

If there are linebreaks, and no lines have trailing or leading spaces,
then use folded or literal.  Prefer literal if all lines are below a
reasonable limit.

If neither literal nor folded are possible, then fall back to
double-quoted.

Fixes #66

isaacs commented:
review, please: @nodeca

The `StringBuilder` class is probably unnecessary, since it turned out that it was simpler to just use booleans for the other dump types.  But I left it in since it's a bit nicer way to organize that logic.  I'm happy to refactor it out if you feel it's not worth it.

puzrin commented:
How about speed change of new dumper? I don't ask about every case, only about situation in general.

/cc @dervus please take a look

isaacs commented:
@puzrin Do you have a benchmark script that it could be tested against?

I try to avoid making optimizations without having some way of measuring specific impact.  It's your project, so I'm happy to polish it as much as you'd like.  But most of the "slow path" things will only start to be slow once you have a very long multi-line string, which we already have buffered in memory, and the use of slice and string appending takes advantage of the cons string capabilities in V8.  Comparing characters vs charcodes is not really a significant difference; it's only 0.39% faster in my micro-benchmarks, on an operation that can be performed at about 10,000 ops/ms on my machine.  A slightly greater gain could be done by simply moving the character code vars inline in the functions that use them, rather than defining them at the module-level.  (Also, those should be `const`s, which will get us some more speed improvements, at the cost of losing older runtimes.  Maybe a TODO for later?)

Much greater gains could be made by tracking down deopts and inline opportunities.

isaacs commented:
Sorry, I was measuring wrong.  Not 10,000 ops/ms.  200,000 ops/ms.

Character and integer comparisons are *crazy fast* in v8!

puzrin commented:
@isaacs there are no separate dumper benchmark (nobody cared much). We have only https://github.com/nodeca/js-yaml/tree/master/benchmark and extend it when needed.

puzrin commented:
@isaacs i've finished my review pass (generic design/logic). LGTM.  It worth to wait @dervus, because he is more experienced in yaml specs. He told me he will look carefully tommorow.

Please, squash into single commit.

dervus commented:
The first thing I noticed is the test. Most likely you want to create individual test for that, since test/samples-common do not check formatting; just values of load and dump-load. Will take a look at the rest tomorrow. Thanks for such massive PR anyway!

dervus commented:
IMO it should be `test/units/dump-scalar-styles.js`

isaacs commented:
@dervus @puzrin Thank you both for the review.  I'm looking forward to any new stuff you find :)

I've updated a few more things, and added a unit test.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -132,6 +132,30 @@ function State(options) {
+  var next = -1;
+  var result = '';
+  var line;
+  var length = string.length;
```
Please use single `var` statement for the whole block of assignments. See any other place in our code for a reference.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -132,6 +132,30 @@ function State(options) {
+      result += ind;
+    }
+    result += line;
+  }
```
Please always wrap code blocks with empty lines.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -132,6 +132,30 @@ function State(options) {
+  }
+  return result;
+}
+
```
Need 2 empty lines between top-level definitions.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
+    max = 40;
+  }
+
+  for (var position = 0; position < object.length; position++) {
```
All `var` statements must be declared at top-level of function body. Also it's preferred, but not required to declare them at very beginning of the function body.

isaacs commented:
@dervus Thanks, I'll update those style points soon.

Aside on that, would it be worthwhile to use a linter that catches these kinds of things as well?  I think jshint and eslint are both pretty popular, and have more options.  Jslint is a bit behind the times feature-wise, I think.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
+    simple = false;
+  }
+
+  var single = true;
```
This is called "plain" in YAML, but if you think "simple" is better, I do not insist.

puzrin commented:
@isaacs i've already migrated most of projects to eslint. This happens when repo activity rises. Will update js-yaml too after your PR megre.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
+    return;
+  }
+
+  var simple = true;
```
This is called "plain" in YAML, but if you think "simple" is better, I do not insist.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
+                   CHAR_SPACE === object.charCodeAt(object.length - 1));
+
+  var maybeNum = (CHAR_MINUS    === first ||
+                  CHAR_QUESTION === first);
```
I don't quite understand the purpose of this.

isaacs replied:
It's a carry-over from the previous dumper implementation that would not dump a string as a simple (unquoted) scalar if it started with `-` or `?`.  I'm not sure why.  I think it is a numeric format thing?

dervus replied:
Ah! Yeah, but it's actually not about numbers. Number-looking strings are secured by the implicit resolving check. It's actually a simplified check for "? " and "- " character sequences that denote mappings and sequences correspondingly. However there should be more of such characters, not only `-` and `?`. There is also `@` at least. See http://www.yaml.org/spec/1.2/spec.html#ns-plain-first%28c%29

isaacs replied:
I think `@` and the other chars will trigger `simple = false`, no matter where they appear in the string.  It seems that only `-` and `?` are ok in a plain scalar as long as they don't appear at the start.

In general, I tried to only add valid folding/literal scalar dumping styles, and not change anything about what was allowed in plain unquoted scalars.

dervus commented on lib/js-yaml/dumper.js:
```diff
@@ -150,81 +174,331 @@ function testImplicitResolving(state, str) {
 
-function writeScalar(state, object) {
-  var isQuoted, checkpoint, position, length, character, first;
+function StringBuilder (source) {
```
Do not put a whitespace after function names.

dervus commented:
Other than style issues it seems okay for me.

isaacs commented:
@dervus That `CHAR_SPACE || CHAR_QUESTION` thing was added in https://github.com/nodeca/js-yaml/commit/b946fcdc6edee8da19ab7bc5c5fa01e9ba03eb40 by @puzrin to close #104.  I'm not sure what cases it was intended to retain quoting for, but I didn't want to change behavior for that situation.

Here's some edge cases that might be relevant:

```
> require('./').safeLoad('---asdf')
'asdf'
> require('./').safeLoad('--asdf')
'--asdf'
> require('./').safeLoad('--- asdf')
'asdf'
> require('./').safeLoad('??? asdf')
'??? asdf'
> require('./').safeLoad('??? --- asdf')
'??? --- asdf'
> require('./').safeLoad('? --- asdf')
{ '--- asdf': null }
```

It seems like quoting things that start with `?` or `-` is prudent, since it might be some magic yaml stuff.  Perhaps I was misunderstanding the intent, though, and the name "maybeNum" is actually the only problem.  I'll rename it to `maybeYamlMagic` or something :)

isaacs commented:
@dervus Simplified the confusing var name, and fixed the style issues you pointed out.  Should be good to go.

dervus commented:
The only magic here is the fact that we just used the simplest and dumbiest solution for quickly solve the problem and go on. :D Here I'll duplicate the link there you can see how it should be handled: http://www.yaml.org/spec/1.2/spec.html#ns-plain-first%28c%29

isaacs commented:
@dervus Ah, ok, that makes a lot of sense.

In that case, yes, there is still an opportunity to do more careful string examination here, if there is a goal to avoid quoting in more cases that could potentially be dumped as plain scalars.  For example, `"a!b"` should be a valid scalar, but as of this pull req will be quoted.

I think we can leave that for the next adventure.  The behavior here is not *incorrect*, and an overall improvement from what was there before.

puzrin commented:
@dervus , merge?

dervus commented:
Yeah, there is no much need for making strings plain by any cost. This PR is about block scalars anyway. So thanks for contribution!

dervus commented:
@isaacs Sorry, I have mislead you about two lines between top-level file statements. Seems like we dropped this style, and I forgot about it. I've corrected that.

puzrin commented:
I've finished other changesets, but suggest to wait 1-2 days before release, if not urgent. May be someone remember more breaking/pending changes.

puzrin commented:
Released 3.3.0. Thanks again for your work!
